### PR TITLE
add augmentation module

### PR DESCRIPTION
**Pull request by ZhaoChuyang**
**WIP**

PR for GSoC'22 project on Efficient Data Augmentation Module in OpenCV for DL Training

I implemented data augmentation methods based on basic image processing and have tested the performance of these already implemented methods in python enviroment. The following table demonstrates the running time (in seconds) comparison between OpenCV-Aug module and torchvision transforms on a subset of ImageNet:

**single method:**

method | config | dataset | opencv | torchvision
-- | -- | -- | -- | --
resize | size: (200, 200) | imagenet-320 | 0.388 | 1.745
center crop | size: (200, 200) | imagenet-320 | 0.001 | 0.017
pad | padding: (100, 100, 100, 100) | imagenet-320 | 0.141 | 0.553
random crop | size: (200, 200) | imagenet-320 | 0.001 | 0.028
random resized crop | size: (500, 500) | imagenet-320 | 0.368 | 3.806
random flip | default | imagenet-320 | 0.046 | 0.519

**compose multiple methods:**

method | config | num_augs | opencv | pytorch
-- | -- | -- | -- | --
RandomCrop + RandomFlip + Pad | RandomCrop: size(300, 300) Pad: padding(100, 100, 100, 100) | 3 | 0.481 | 1.240
Resize + Pad + RandomFlip + CenterCrop | Resize: size(400, 400) Pad:  padding(100, 100, 100, 100) CenterCrop: size(200, 200) | 5 | 0.636 | 5.486

To further improve the performance of augmentation methods, all operations on images are excuted in-place if possible.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
This patch should be targeted to **opencv_contrib** repository.

**Comment by asmorkalov** [comment]
@ZhaoChuyang Thanks for the contribution! The PR and GSoC project was discussed on OpenCV Core Team meeting and the team proposed to migrate the module to OpenCV Contrib for now. Please close the PR and propose the new one to [opencv_contrib](https://github.com/opencv/opencv_contrib) repo. Reason: we want stabilize API first and m.b. add more tests. API changes after public release will be not possible or very hard. We do not have strict API compatibility in contrib for now.

**Comment by ZhaoChuyang** [comment]
OK I get it. I'll close this PR and open a new one in the contrib repo.
